suite/tests/client-interface/drreg-end-restore.dll.c

```diff
@@ -0,0 +1,194 @@
+
+DR_EXPORT void
+dr_init(client_id_t id)
+{
+    drreg_options_t ops = { sizeof(ops), 2 /*max slots needed*/, false };
+    drmgr_priority_t priority = { sizeof(priority), "my_priority", NULL, NULL, 0 };
+    bool ok;
+    drreg_status_t res;
+
+    dr_set_client_name("DynamoRIO Sample Client 'drreg-end-restor'",
+                       "http://dynamorio.org/issues");
+
+    drmgr_init();
+    res = drreg_init(&ops);
+    CHECK(res == DRREG_SUCCESS, "drreg init failed");
+    dr_register_exit_event(event_exit);
+
+    ok = drmgr_register_bb_instrumentation_ex_event(event_bb_app2app, event_bb_analysis,
+                                                    event_bb_insert, NULL, &priority);
+    CHECK(ok, "drmgr register bb failed");
+
+    tls_idx = drmgr_register_tls_field();
+    CHECK(tls_idx != -1, "tls registration failed");
+}
+
+static void
+event_exit(void)
+{
+    drmgr_unregister_tls_field(tls_idx);
+
+    bool ok = drmgr_unregister_bb_instrumentation_ex_event(
+        event_bb_app2app, event_bb_analysis, event_bb_insert, NULL);
+    CHECK(ok, "drmgr unregister bb failed");
+    CHECK(performed_check, "check was not performed");
+    drreg_exit();
+    drmgr_exit();
+}
+
+static dr_emit_flags_t
+event_bb_app2app(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                 bool translating, OUT void **user_data)
+{
+    drreg_status_t res = drreg_set_bb_properties(drcontext, DRREG_IGNORE_BB_END_RESTORE);
+    CHECK(res == DRREG_SUCCESS, "failed to set property");
+    return DR_EMIT_DEFAULT;
+}
+
+static dr_emit_flags_t
+event_bb_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                  bool translating, void *user_data)
+{
+    drreg_status_t res = drreg_set_bb_properties(drcontext, DRREG_IGNORE_BB_END_RESTORE);
+    CHECK(res == DRREG_SUCCESS, "failed to set property");
+    return DR_EMIT_DEFAULT;
+}
+
+static reg_t
+get_val_from_ctx(void *drcontext, reg_id_t reg_id)
+{
+    dr_mcontext_t mcontext = {
+        sizeof(mcontext),
+        DR_MC_ALL,
+    };
+    dr_get_mcontext(drcontext, &mcontext);
+    return reg_get_value(reg_id, &mcontext);
+}
+
+static void
+set_reg_val()
+{
+    void *drcontext = dr_get_current_drcontext();
+    void *val = (void *)get_val_from_ctx(drcontext, TEST_REG);
+    drmgr_set_tls_field(drcontext, tls_idx, val);
+}
+
+static void
+check_reg_val()
+{
+    void *drcontext = dr_get_current_drcontext();
+    reg_t val = get_val_from_ctx(drcontext, TEST_REG);
+    reg_t orig_val = (reg_t)drmgr_get_tls_field(drcontext, tls_idx);
+    CHECK(val == orig_val, "restoration failed");
+    performed_check = true;
+}
+
+static dr_emit_flags_t
+event_bb_insert(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
+                bool for_trace, bool translating, void *user_data)
+{
+    drreg_status_t res;
+    reg_id_t reg;
+    drvector_t allowed;
+    bool is_dead;
+
+    drreg_is_register_dead(drcontext, TEST_REG, instr, &is_dead);
```

Comment:
Missing check of return status here and on several drreg routines below.

Reply:
Done